uvicorn/server.py

```diff
@@ -302,22 +307,28 @@ async def _wait_tasks_to_complete(self) -> None:
         for server in self.servers:
             await server.wait_closed()
 
-    def install_signal_handlers(self) -> None:
+    @contextlib.contextmanager
+    def capture_signals(self) -> Generator[None, None, None]:
+        # Signals can only be listened to from the main thread.
         if threading.current_thread() is not threading.main_thread():
-            # Signals can only be listened to from the main thread.
+            yield
             return
-
-        loop = asyncio.get_event_loop()
-
+        # always use signal.signal, even if loop.add_signal_handler is available
+        # this allows to restore previous signal handlers later on
+        original_handlers = {sig: signal.signal(sig, self.handle_exit) for sig in HANDLED_SIGNALS}
```

Comment:
Why did we use `loop.add_signal_handler` before?

Reply:
I haven't found a technical reason for it. It seems to have been added in #141 for no specific and  then copied over the years. I assume it was used because the `asyncio` docs promote it as a Linux feature that is better in unspecified ways.

On the technical side the *only* advantage of `loop.add_signal_handler` is that it allows synchronously triggering async code (e.g. `event.set()`), and the handlers don't use that. Since the code has to be compatible with Windows, handlers cannot rely on being invoked by the loop anyways.